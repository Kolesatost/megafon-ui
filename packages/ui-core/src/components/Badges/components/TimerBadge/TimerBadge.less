@import (reference) '~styles/base.less';

@b: e('.mfui-timer-badge');

@{b} {
    display: flex;
    align-items: center;

    padding: 1px 2px;
    padding-right: 8px;
    width: fit-content;
    height: fit-content;

    border: 1px solid var(--spbSky1);
    border-radius: 12px;

    background: var(--base);
    overflow: hidden;
    
    &_theme_grey {
        @{b}__text {
            color: var(--spbSky3);
        }

        @{b}__icon {
            fill: var(--spbSky3);
        }
    }

    &_theme_red {
        @{b}__text {
            color: var(--fury);
        }

        @{b}__icon {
            fill: var(--fury);
        }
    }

    &__text {
        .smallFont();
        .fontFamily();

        margin-left: 2px;

        font-weight: 500;
    }

    &__icon {
        height: 20px;
        width: 20px;
    }

    &__icon-container {
        position: relative;
        display: flex;
    }

    &__icon-container_shadow:before {
        content: '';
        position: absolute;
        top: 2px;
        left: 2px;

        height: 16px;
        width: 16px;

        border-radius: 50%;
        box-shadow: 0px 0px 6px var(--fury);
    }
}
