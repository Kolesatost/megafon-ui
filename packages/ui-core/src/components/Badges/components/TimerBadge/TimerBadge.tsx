import React, { useEffect, useState, useRef } from 'react';
import { cnCreate, filterDataAttrs } from '@megafon/ui-helpers';
import TimerIcon from '@megafon/ui-icons/system-16-timer_16.svg';
import * as PropTypes from 'prop-types';
import Timer from './Timer/Timer';
import './TimerBadge.less';

const SECONDS_IN_HOUR = 3600;
const MS_IN_SECOND = 1000;
const TIMEOUT = 1000;

const TimerBadgeTheme = {
    RED: 'red',
    GREY: 'grey',
} as const;

export interface ITimerBadgeProps {
    /** Дата окончания таймера */
    expirationDate: Date;
    /** Время начала обратного отсчёта, в секундах */
    countdownStart?: number;
    /** Отображение дополнительного текста */
    hasPrefix?: boolean;
    /** Дополнительный текст для обратного отсчёта */
    countdownText?: string;
    /** Дополнительный текст для даты окончания таймера */
    expirationDateText?: string;
    /** Дополнительный класс корневого элемента */
    className?: string;
    /** Дополнительные data-атрибуты к внутренним элементам */
    dataAttrs?: {
        root?: Record<string, string>;
        text?: Record<string, string>;
        timer?: Record<string, string>;
    };
}

const cn = cnCreate('mfui-timer-badge');
const TimerBadge: React.FC<ITimerBadgeProps> = ({
    expirationDate,
    countdownStart = 86400,
    hasPrefix = false,
    countdownText = 'Закончится через',
    expirationDateText = 'Воспользуйтесь до',
    className,
    dataAttrs,
}) => {
    const countTimeout = useRef<number>(0);

    const remainingTime = (new Date(expirationDate).getTime() - new Date().getTime()) / MS_IN_SECOND;

    const [actualRemainingTime, setActualRemainingTime] = useState(remainingTime);
    const [showCountdown, setShowCountdown] = useState(false);

    const isLastHour = actualRemainingTime < SECONDS_IN_HOUR;
    const isTimeExpired = actualRemainingTime < 1 || new Date(expirationDate) <= new Date();

    const theme = showCountdown ? TimerBadgeTheme.RED : TimerBadgeTheme.GREY;
    const additionalText = showCountdown ? countdownText : expirationDateText;

    useEffect(() => {
        setShowCountdown(actualRemainingTime < countdownStart);
    }, [actualRemainingTime, countdownStart]);

    useEffect(() => {
        if (isTimeExpired) {
            return undefined;
        }

        countTimeout.current = window.setTimeout(() => setActualRemainingTime(actualRemainingTime - 1), TIMEOUT);

        return (): void => clearTimeout(countTimeout.current);
    }, [isTimeExpired, showCountdown, actualRemainingTime]);

    return (
        <div {...filterDataAttrs(dataAttrs?.root)} className={cn({ theme }, className)}>
            <div className={cn('icon-container', { shadow: isLastHour && showCountdown })}>
                <TimerIcon className={cn('icon')} />
            </div>
            <div {...filterDataAttrs(dataAttrs?.text)} className={cn('text')}>
                {isTimeExpired ? (
                    'Время действия истекло'
                ) : (
                    <Timer
                        additionalText={hasPrefix ? additionalText : undefined}
                        showCountdown={showCountdown}
                        expirationDate={expirationDate}
                        actualRemainingTime={actualRemainingTime}
                        dataAttrs={{ root: dataAttrs?.timer }}
                    />
                )}
            </div>
        </div>
    );
};

TimerBadge.propTypes = {
    expirationDate: PropTypes.instanceOf(Date).isRequired,
    countdownStart: PropTypes.number,
    hasPrefix: PropTypes.bool,
    countdownText: PropTypes.string,
    expirationDateText: PropTypes.string,
    className: PropTypes.string,
    dataAttrs: PropTypes.shape({
        root: PropTypes.objectOf(PropTypes.string.isRequired),
        text: PropTypes.objectOf(PropTypes.string.isRequired),
        timer: PropTypes.objectOf(PropTypes.string.isRequired),
    }),
};

export default TimerBadge;
