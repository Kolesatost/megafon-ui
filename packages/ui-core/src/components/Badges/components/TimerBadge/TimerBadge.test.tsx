import * as React from 'react';
import { mount } from 'enzyme';
import TimerBadge from './TimerBadge';

const RealDate = global.Date;
const date = new Date('2022-03-15T10:50:15z');

const defaultDate = new Date('2022-04-25T14:50:15z');

const someHoursBeforeEnd = new Date('2022-03-15T15:50:15z');
const lastHour = new Date('2022-03-15T11:20:15z');
const expiredDate = new Date('2022-03-15T09:50:15z');

const defaultAllProps = {
    dataAttrs: {
        root: { 'data-testid': 'root' },
        text: { 'data-testid': 'text' },
        timer: { 'data-testid': 'timer' },
    },
    className: 'rootClass',
    expirationDate: defaultDate,
    countdownStart: 604800,
    countdownText: 'Срок действия истечёт через',
    expirationDateText: 'Действительно до',
    hasPrefix: true,
};

describe('<TimerBadge />', () => {
    describe('snapshots', () => {
        beforeAll(() => {
            global.Date = class extends Date {
                constructor(customDate: string) {
                    super();

                    return customDate ? new RealDate(customDate) : date;
                }
            } as DateConstructor;
        });

        afterAll(() => {
            global.Date = RealDate;
        });

        it('renders with require props', () => {
            const wrapper = mount(<TimerBadge expirationDate={defaultDate} />);

            expect(wrapper).toMatchSnapshot();
        });

        it('renders with all props', () => {
            const wrapper = mount(<TimerBadge {...defaultAllProps} />);

            expect(wrapper).toMatchSnapshot();
        });

        it('renders with countdown', () => {
            const wrapper = mount(<TimerBadge expirationDate={someHoursBeforeEnd} hasPrefix />);

            expect(wrapper).toMatchSnapshot();
        });

        it('renders with last hour', () => {
            const wrapper = mount(<TimerBadge expirationDate={lastHour} hasPrefix />);

            expect(wrapper).toMatchSnapshot();
        });

        it('renders with expired time', () => {
            const wrapper = mount(<TimerBadge expirationDate={expiredDate} />);

            expect(wrapper).toMatchSnapshot();
        });
    });
});
