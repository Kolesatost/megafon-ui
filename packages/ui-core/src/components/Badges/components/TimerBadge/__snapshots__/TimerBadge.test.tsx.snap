// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimerBadge /> snapshots renders with all props 1`] = `
<TimerBadge
  className="rootClass"
  countdownStart={604800}
  countdownText="Срок действия истечёт через"
  dataAttrs={
    Object {
      "root": Object {
        "data-testid": "root",
      },
      "text": Object {
        "data-testid": "text",
      },
      "timer": Object {
        "data-testid": "timer",
      },
    }
  }
  expirationDate={2022-04-25T14:50:15.000Z}
  expirationDateText="Действительно до"
  hasPrefix={true}
>
  <div
    className="mfui-timer-badge mfui-timer-badge_theme_grey rootClass"
    data-testid="root"
  >
    <div
      className="mfui-timer-badge__icon-container"
    >
      <test-file-stub
        className="mfui-timer-badge__icon"
      />
    </div>
    <div
      className="mfui-timer-badge__text"
      data-testid="text"
    >
      <Timer
        actualRemainingTime={3556800}
        additionalText="Действительно до"
        dataAttrs={
          Object {
            "root": Object {
              "data-testid": "timer",
            },
          }
        }
        expirationDate={2022-04-25T14:50:15.000Z}
        showCountdown={false}
      >
        <div
          className="mfui-timer"
          data-testid="timer"
        >
          <span>
            Действительно до
             
          </span>
          <span>
            25 апреля 2022
          </span>
        </div>
      </Timer>
    </div>
  </div>
</TimerBadge>
`;

exports[`<TimerBadge /> snapshots renders with countdown 1`] = `
<TimerBadge
  expirationDate={2022-03-15T15:50:15.000Z}
  hasPrefix={true}
>
  <div
    className="mfui-timer-badge mfui-timer-badge_theme_red"
  >
    <div
      className="mfui-timer-badge__icon-container"
    >
      <test-file-stub
        className="mfui-timer-badge__icon"
      />
    </div>
    <div
      className="mfui-timer-badge__text"
    >
      <Timer
        actualRemainingTime={18000}
        additionalText="Закончится через"
        dataAttrs={
          Object {
            "root": undefined,
          }
        }
        expirationDate={2022-03-15T15:50:15.000Z}
        showCountdown={true}
      >
        <div
          className="mfui-timer"
        >
          <span>
            Закончится через
             
          </span>
          <span>
            5 ч
          </span>
        </div>
      </Timer>
    </div>
  </div>
</TimerBadge>
`;

exports[`<TimerBadge /> snapshots renders with expired time 1`] = `
<TimerBadge
  expirationDate={2022-03-15T09:50:15.000Z}
>
  <div
    className="mfui-timer-badge mfui-timer-badge_theme_red"
  >
    <div
      className="mfui-timer-badge__icon-container mfui-timer-badge__icon-container_shadow"
    >
      <test-file-stub
        className="mfui-timer-badge__icon"
      />
    </div>
    <div
      className="mfui-timer-badge__text"
    >
      Время действия истекло
    </div>
  </div>
</TimerBadge>
`;

exports[`<TimerBadge /> snapshots renders with last hour 1`] = `
<TimerBadge
  expirationDate={2022-03-15T11:20:15.000Z}
  hasPrefix={true}
>
  <div
    className="mfui-timer-badge mfui-timer-badge_theme_red"
  >
    <div
      className="mfui-timer-badge__icon-container mfui-timer-badge__icon-container_shadow"
    >
      <test-file-stub
        className="mfui-timer-badge__icon"
      />
    </div>
    <div
      className="mfui-timer-badge__text"
    >
      <Timer
        actualRemainingTime={1800}
        additionalText="Закончится через"
        dataAttrs={
          Object {
            "root": undefined,
          }
        }
        expirationDate={2022-03-15T11:20:15.000Z}
        showCountdown={true}
      >
        <div
          className="mfui-timer"
        >
          <span>
            Закончится через
             
          </span>
          <span>
            30 мин
          </span>
        </div>
      </Timer>
    </div>
  </div>
</TimerBadge>
`;

exports[`<TimerBadge /> snapshots renders with require props 1`] = `
<TimerBadge
  expirationDate={2022-04-25T14:50:15.000Z}
>
  <div
    className="mfui-timer-badge mfui-timer-badge_theme_grey"
  >
    <div
      className="mfui-timer-badge__icon-container"
    >
      <test-file-stub
        className="mfui-timer-badge__icon"
      />
    </div>
    <div
      className="mfui-timer-badge__text"
    >
      <Timer
        actualRemainingTime={3556800}
        dataAttrs={
          Object {
            "root": undefined,
          }
        }
        expirationDate={2022-04-25T14:50:15.000Z}
        showCountdown={false}
      >
        <div
          className="mfui-timer"
        >
          <span>
            25 апреля 2022
          </span>
        </div>
      </Timer>
    </div>
  </div>
</TimerBadge>
`;
